### ISSUE TITLE
Issues generating multiple schema formats when decorating SchemaFactory

### ISSUE THREAD

**Issue by GwendolenLynch** [api-platform/core#6228]
**API Platform version(s) affected**: 3.2

**Description**  
I am working on a PR to introduce a JSON:API schema factory and hit a couple of related bugs that feel better suited to a separate issue/PR combination.

I have working "fixes" laid out below, but would like them reviewed & sanity checked before I create a PR for this.

First problem was that `ApiPlatform\Hydra\JsonSchema\SchemaFactory#addDistinctFormat()` checks for `ApiPlatform\JsonSchema\SchemaFactory`, where as `ApiPlatform\Hal\JsonSchema\SchemaFactory` simply does a `method_exists` call.

No `priority` is set on any of the service decorators, and Hydra's, by chance, is currently the one with the base service  `.inner`. However, my new service took first place and it became impossible to build both schema formats at the same time as the last one "won out". Changing Hydra's method behaviour to be a `method_exists` call fixes this.

The second is just as subtle. The base `ApiPlatform\JsonSchema\SchemaFactory` calls the injected decorator service to enable subschema generation (see #6055). When the Hydra service is the one decorating the base this works fine, but the introduction of another service causes that one to be called instead, meaning Hydra misses out (see #5950)

**How to reproduce**  

```php
final class SchemaFactory implements SchemaFactoryInterface
{
    public function __construct(private readonly SchemaFactoryInterface $schemaFactory)
    {
        $this->addDistinctFormat('jsonapi');
        if ($this->schemaFactory instanceof SchemaFactoryAwareInterface) {
            $this->schemaFactory->setSchemaFactory($this);
        }
    }

    public function buildSchema(string $className, string $format = 'jsonapi', string $type = Schema::TYPE_OUTPUT, ?Operation $operation = null, ?Schema $schema = null, ?array $serializerContext = null, bool $forceCollection = false): Schema
    {
        $schema = $this->schemaFactory->buildSchema($className, $format, $type, $operation, $schema, $serializerContext, $forceCollection);

        if ('jsonapi' !== $format) {
            return $schema;
        }

        // Do schema updates here …

        return $schema;
    }

    public function addDistinctFormat(string $format): void
    {
        // The Hydra schema factory does:
        //  if ($this->schemaFactory instanceof SchemaFactoryAwareInterface) {
        //
        // The HAL one does as below:
        if (method_exists($this->schemaFactory, 'addDistinctFormat')) {
            $this->schemaFactory->addDistinctFormat($format);
        }
    }
}
```

Obligatory service definition :innocent: 
```yaml
    api_platform.jsonapi.json_schema.schema_factory:
        class: 'ApiPlatform\JsonApi\JsonSchema\SchemaFactory'
        decorates: 'api_platform.json_schema.schema_factory'
        arguments:
            - '@api_platform.jsonapi.json_schema.schema_factory.inner'
```

And API Platform config:
```
api_platform:
    formats:
        json: [ 'application/json' ]
        jsonld: [ 'application/ld+json' ]
        jsonhal: [ 'application/hal+json' ]
        jsonapi: [ 'application/vnd.api+json' ]
```

**Possible Solution**  
1. Change Hydra `SchemaFactory#addDistinctFormat()` to use a `method_exists` call
2. Have all `SchemaFactory`s implement `SchemaFactoryAwareInterface` :disappointed: 

Obviously neither of those are particularly elegant, hence this discussion. Help and guidance most welcome!

**Comment by GwendolenLynch** [comment]
With respect to the `SchemaFactory#addDistinctFormat()` problem, changing these lines in the fixture app's config:
https://github.com/api-platform/core/blob/1cc114b93e95bdce5b48be46f84c1e11446a8d2f/tests/Fixtures/app/config/config_common.yml#L36-L39

To something like this:
```yaml
    formats:
        jsonapi:                       ['application/vnd.api+json']
        jsonhal:                       ['application/hal+json']
        jsonld:                        ['application/ld+json']
```

Then running `vendor/bin/simple-phpunit --filter JsonSchemaGenerateCommandTest::` will make that one trigger, and can be fixed with the `method_exists` suggestion above.

I'm trying to work out how to add specific tests for this and then I'll PR Suggestions gratefully welcomed.

**Comment by GwendolenLynch** [comment]
> > The second is just as subtle. The base ApiPlatform\JsonSchema\SchemaFactory calls the injected decorator service to enable subschema generation (see https://github.com/api-platform/core/pull/6055). When the Hydra service is the one decorating the base this works fine, but the introduction of another service causes that one to be called instead, meaning Hydra misses out (see https://github.com/api-platform/core/issues/5950)
> 
> Not sure I got that, the whole decoration chain should be called even in the subschema no?

Yes, it should. What I am observing happen is that the decoration chain I had was HAL -> Hydra -> JSON:API -> JSON, and when JSON would call the sub-schema factory it would short circuit to I think Hydra, missing the others. It all came down to what order the `formats:` key was in.

**Comment by soyuka** [comment]
Lmk if you can reproduce that as I don't understand the link between the formats order and the decoration call order.

**Comment by GwendolenLynch** [comment]
With your suggested changes above I can't reproduce it (decoration call order). I've updated #6236 to incorporate your idea, and rollback the `implements` changes.

Currently, I have left empty `addDistinctFormat()` methods. The base one had an `@internal` annotation, the others not, so I'd appreciate feedback on remove/deprecate decision.

**Comment by GwendolenLynch** [comment]
Fixed in #6236
